Enable Cross-Origin requests on /api/v1

hyleung opened the pull request:
Closes #1135 (unless, of course, I totally misunderstood the issue 😃)

hyleung committed:
Add ability to configure CORS for zipkin-server

This commit enables CORS on /api/v1/ *by default*.

allowed-origins can be configured via the Spring Boot config file (e.g. `zipkin-server.yml`).

e.g.

zipkin:
   ...
  query:
    # CORS allowed-origins.
    allowed-origins: "*"

hyleung commented:
Hmmm... not sure why there's a test failure. I ran `./mvnw  clean install` locally and it seems to build ok.

codefromthecrypt commented:
Is `@CrossOrigin(*)` the way to do this, or is there a parameterized way to do this in config that would allow it to be tightened? cc @shakuzen @dsyer

hyleung commented:
Might be able to do it in a `ConfigurerAdapter`. I'll have a look into it!
On Wed, Aug 10, 2016 at 6:57 PM Adrian Cole notifications@github.com
wrote:

> Is @CrossOrigin(*) the way to do this, or is there a parameterized way to
> do this in config that would allow it to be tightened? cc @shakuzen
> https://github.com/shakuzen @dsyer https://github.com/dsyer
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/openzipkin/zipkin/pull/1234#issuecomment-239055450,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAPH-v4J9ZLCfsfTPrv2-AeUA7noqXB8ks5qeoGGgaJpZM4JhLqg
> .

hyleung commented:
Looking at the [docs](https://spring.io/guides/gs/rest-service-cors/), think you can do it like this:

```
@Bean
public WebMvcConfigurer corsConfigurer() {
    return new WebMvcConfigurerAdapter() {
        @Override
        public void addCorsMappings(CorsRegistry registry) {
            registry.addMapping("/greeting-javaconfig").allowedOrigins("http://localhost:9000");
        }
    };
}
```

Maybe we could get the list of allowed origins out of `zipkin-server.yml`? Something like:

```
server:
  port: ${QUERY_PORT:9411}
  compression:
    enabled: true
    # compresses any response over min-response-size (default is 2KiB)
    # Includes dynamic json content and large static assets from zipkin-ui
    mime-types: application/json,application/javascript,text/css,image/svg
  cors:
    -
      path: "/api/v1/*"
      allowed-origins:
        - "*"
        - http://myserver.com
```

shakuzen commented:
First of all, thank you for the PR @hyleung!

As of Spring 4.3, the `@CrossOrigin`'s `origins` parameter supports placeholders ([SPR-14010](https://jira.spring.io/browse/SPR-14010)). Zipkin is already on Spring 4.3.x, and thus, we could just put the annotation on the `ZipkinQueryApiV1` controller with some property like `zipkin.query.cors-origins` to configure the origins, with a default of `*` in `zipkin-server.yml`. This way people can customize it.

It feels a little weird to me to be able to configure the allowed CORS origins for any path on the Zipkin server via properties, as the implementation at the time of this comment allows. I don't think we want to allow that, actually.

hyleung commented:
@shakuzen, awesome, I had no idea you could do that. That's way easier. I'll make some changes and push the branch again at some point today. Thanks!

codefromthecrypt commented:
any news here? will probably cut a release in the next day or so

hyleung commented:
@adriancole I updated the PR based on @shakuzen's feedback - but still can't get the PR build to pass. 

```
Tests in error: 
  multithreaded(zipkin.storage.cassandra.DeduplicatingExecutorTest): test timed out after 2000 milliseconds
```

As far as I can tell, the build is ok locally - maybe there's something else I should check? (I'm just running `./mvnw clean install`)

shakuzen commented on zipkin-server/src/main/resources/zipkin-server.yml:
```diff
@@ -83,13 +83,18 @@ zipkin:
     # - .*example2.com
     # Default is "match all websites"
     instrumented: .*
+  server:
```
I thought reusing the `zipkin.query` namespace would be appropriate for this property. @adriancole WDYT?

hyleung replied:
I suppose of it's only the query API that this applies to (which is the case at the moment) then this would make sense (reusing `zipkin.query`.

codefromthecrypt replied:
agreed. we should re-use zipkin.query as I don't expect us to apply this to other components such as the /health endpoint

shakuzen commented on zipkin-server/src/main/java/zipkin/server/ZipkinQueryApiV1.java:
```diff
@@ -43,6 +44,7 @@
  */
 @RestController
 @RequestMapping("/api/v1")
+@CrossOrigin("${zipkin.server.allowed-origins:''}")
```
Making the default value here also `*` would be a good idea for those making custom Zipkin servers with `@EnableZipkinServer` or those using their own YAML file of properties that might not contain the new property.
That is, if we want the default value to be `*` in these cases also. Do we? Or do we want people in those situations to make a conscious decision about the allowed origins for the query API?

codefromthecrypt replied:
Provided we document cross origin in the zipkin-server/README I'm cool with
defaulting to *

Ex. a section similar to Self-Tracing

On Tue, Aug 16, 2016 at 9:58 PM, Tommy Ludwig notifications@github.com
wrote:

> In zipkin-server/src/main/java/zipkin/server/ZipkinQueryApiV1.java
> https://github.com/openzipkin/zipkin/pull/1234#discussion_r74939322:
> 
> > @@ -43,6 +44,7 @@
> >   */
> >  @RestController
> >  @RequestMapping("/api/v1")
> > +@CrossOrigin("${zipkin.server.allowed-origins:''}")
> 
> Making the default value here also \* would be a good idea for those
> making custom Zipkin servers with @EnableZipkinServer or those using
> their own YAML file of properties that might not contain the new property.
> That is, if we want the default value to be \* in these cases also. Do we?
> Or do we want people in those situations to make a conscious decision about
> the allowed origins for the query API?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/openzipkin/zipkin/pull/1234/files/607ccc3512ff70c8d6dd946bca7903e0730f96d3#r74939322,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAD618AuBfDqSbM-nHOKkOWf-bQVRfqpks5qgcH6gaJpZM4JhLqg
> .

codefromthecrypt commented:
rebase against latest master and you should be ok on this test failure

On Tue, Aug 16, 2016 at 9:43 PM, Ho Yan Leung notifications@github.com
wrote:

> @adriancole https://github.com/adriancole I updated the PR based on
> @shakuzen https://github.com/shakuzen's feedback - but still can't get
> the PR build to pass.
> 
> Tests in error:
>   multithreaded(zipkin.storage.cassandra.DeduplicatingExecutorTest): test timed out after 2000 milliseconds
> 
> As far as I can tell, the build is ok locally - maybe there's something
> else I should check? (I'm just running ./mvnw clean install)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/openzipkin/zipkin/pull/1234#issuecomment-240105428,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAD6199FMeya8wDTSqJL_2jCelZ_LEmfks5qgb56gaJpZM4JhLqg
> .

hyleung commented:
@adriancole hmm...I rebased, pretty sure at the moment I'm up to date vs. upstream getting:

```
Tests in error: 
  zipkin.collector.kafka.KafkaCollectorTest: test timed out after 10 seconds
```

In any case, I've made the changes discussed above and added a little blurb to the README

codefromthecrypt commented:
Travis has been more flakey lately on heavy tests. I will kick it

On 17 Aug 2016 3:42 am, "Ho Yan Leung" notifications@github.com wrote:

> @adriancole https://github.com/adriancole hmm...I rebased, pretty sure
> at the moment I'm up to date vs. upstream getting:
> 
> Tests in error:
>   zipkin.collector.kafka.KafkaCollectorTest: test timed out after 10 seconds
> 
> In any case, I've made the changes discussed above and adding a little
> blurb to the README
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/openzipkin/zipkin/pull/1234#issuecomment-240214751,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAD61-_8FH0PQEF4nITklM39PH7bl9U4ks5qghKIgaJpZM4JhLqg
> .

shakuzen commented on zipkin-server/src/main/java/zipkin/server/ZipkinQueryApiV1.java:
```diff
@@ -43,6 +44,7 @@
  */
 @RestController
 @RequestMapping("/api/v1")
+@CrossOrigin("${zipkin.query.allowed-origins:'*'}")
```
I don't know if these single quotes around the `*` are needed or make a difference.
A few UT/IT to verify the situations we are covering would, of course, be best. Manual tests in the meantime might suffice.

hyleung replied:
Yeah the single quotes aren't needed - removed them. 

Happy to try and get some ITs in there - are the ones in `zipkin-server/src/it` a good example to start with?

codefromthecrypt replied:
yeah src/it would be ideal!

On Wed, Aug 17, 2016 at 9:03 AM, Ho Yan Leung notifications@github.com
wrote:

> In zipkin-server/src/main/java/zipkin/server/ZipkinQueryApiV1.java
> https://github.com/openzipkin/zipkin/pull/1234#discussion_r75048464:
> 
> > @@ -43,6 +44,7 @@
> >   _/
> >  @RestController
> >  @RequestMapping("/api/v1")
> > +@CrossOrigin("${zipkin.query.allowed-origins:'_'}")
> 
> Yeah the single quotes aren't needed - removed them.
> 
> Happy to try and get some ITs in there - are the ones in
> zipkin-server/src/it a good example to start with?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/openzipkin/zipkin/pull/1234/files/87ecb3a27cd70ad4ec42cc432532cee40f561a82#r75048464,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAD619LSQdcAAAtv2WfFKbQccJGabT2xks5qgl37gaJpZM4JhLqg
> .

codefromthecrypt commented:
@hyleung ps src/it I suppose is optional as long as you've a test in ZipkinServerIntegrationTests

codefromthecrypt commented on zipkin-server/src/main/resources/zipkin-server.yml:
```diff
@@ -29,6 +29,9 @@ zipkin:
     # The Cache-Control max-age (seconds) for /api/v1/services and /api/v1/spans
     names-max-age: 300
+    # CORS allowed-origins.
+    allowed-origins: "*"
```
nit: is the quote needed?

hyleung replied:
I don't think that's considered valid yaml(?) IIRC Spring Boot doesn't like it. Yamllint doesn't like it either...

codefromthecrypt replied:
ah ok. I just saw `.*` used later in the file, so figured it might be valid to use `*`

hyleung replied:
¯_(ツ)_/¯ if it's possible for somebody to "suck at yaml", that would be me. Thinking about it a little more, this entry in the config file is a little redundant now since the default value is \* now - wasn't sure if it was perhaps better "documentation" to leave it in.

codefromthecrypt replied:
we've been adding these... I think it is helpful "documentation"

hyleung commented on zipkin-server/src/it/minimal-dependencies/src/test/java/zipkin/minimal/ZipkinServerCORSTest.java:
```diff
@@ -0,0 +1,69 @@
+@RunWith(SpringJUnit4ClassRunner.class)
+@WebAppConfiguration
+@TestPropertySource(properties = {"zipkin.storage.type=mem", "spring.config.name=zipkin-server", "zipkin.query.allowed-origins=foo.example.com"})
+public class ZipkinServerCORSTest {
```
Originally I wanted to add these tests as part of the fixture that's already there - I thought I'd be able to alter the configuration using `EnvironmentTestUtils.addEnvironment(...)` on a _per test_ basis, but it didn't seem to pick up the updated config. Could very well be my lack of experience with Spring, so if there's a better way, I'm happy to revisit this.

codefromthecrypt replied:
this approach looks great. only thing I'd ask is to move it to the normal src/test tree as it doesn't have anything that requires it to be tested as an integration test. Moving this to normal src/test will make it easier to troubleshoot if it fails.

codefromthecrypt commented on zipkin-server/README.md:
```diff
@@ -17,6 +17,10 @@ The following endpoints are defined for Zipkin:
 
+### CORS (Cross-origin Resource Sharing)
+
+By default, all end points under `/api/v1` are configured to **allow** cross-origin requests. This can be changed by setting the `zipkin.query.allowed-origins` property.
```
last nits.. we are using "endpoint" not "end point". Then put an example, so people know what the value should be. something like...

For example, to only allow requests from mycompany.com, you can set the environment variable `ZIPKIN_QUERY_ALLOWED_ORIGINS=mycompany.com`.

(pretty sure that's the syntax when set invoked as an env variable)

codefromthecrypt commented:
very good. pretty much ready to merge! If you aren't in a position to easy address the last nits, I can do them on the way in. say the word!

hyleung commented:
Cool. Should be able to get those last couple of things done at some point tonight

codefromthecrypt commented:
nice work!

codefromthecrypt commented:
@shakuzen thanks for the review!
